File: leet-code/removeKdigits.js
/**
 * @param {string} num
 * @param {number} k
 * @return {string}
 */
var removeKdigits = function (num, k) {
  const len = num.length
  const stack = []
  const remains = len - k
  for (let i = 0; i < len; i++) {
    while (stack.length > 0 && k > 0 && stack[stack.length - 1] > num[i]) {
      stack.pop()
      k--
    }
    stack.push(num[i])
  }
  while(stack[0]==0){
    stack.shift()
  }
  return stack.slice(0, remains).join('') || '0'
}

console.log(removeKdigits('10', 2))

File: leet-code/removeKdigits.go
package main

import "strings"

func removeKdigits(num string, k int) string {
	numlen := len(num)
	remains := numlen - k
	var stack []byte

	for i := range num {
		for k > 0 && len(stack) > 0 && stack[len(stack)-1] > num[i] {
			stack = stack[:len(stack)-1]
			k--
		}
		stack = append(stack, num[i])
	}
	stack = stack[0:remains]

	res := strings.TrimLeft(string(stack), "0")
	if res == "" {
		res = "0"
	}
	return res
}
